### PR TITLE
architectural overview documentation

### PR DESCRIPTION
**Pull request by peterbe**
Part of #1

I wrote this quickly and it's pretty broad and rough but I think this is how I'd explain it to another human who'd ask me "What the heck is this?"

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,6 +8,62 @@ a threshold.
...
 trigger alerts/notifications that the Product Delivery team can take heed of.
 
+## Architectural Overview
+
+### In Plain English
+
+This is stateless script, written in Python, meant to be executed roughly once a day.
+It uses `requests.get()` to query [Redash](https://sql.telemetry.mozilla.org) for
+a query's data that is pre-emptively run every 24 hours. The data it analyses is a
+list of every collection stored in Remote Settings (Kinto) along with a count of
```
Comment (leplatrem) [inline README.md]: Actually, it targets all the components using the unified Telemetry https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/uptake.html   Remote Settings is one of them. 

Since the default config filters out the other components, I don't know if we should mention that

File: README.md [hunk 2]
```diff
@@ -8,6 +8,62 @@ a threshold.
...
+
+### Historical Strategy
+
+As of Feb 2019, Firefox clients that use Remote Settings only send Telemetry events
+in daily batches. I.e. it's _not_ real-time. The "uptake histogram" is buffered in
+the browser and will send periodically instead of as soon as possible. In Firefox
+Nightly (67 at the time of writing), we are switching to real-time Telemetry Events.
+
+On the Telemetry backend we're still consuming the older uptake histogram but once,
+what is Nightly today, becomes Release, we will switch the Redash query (where
```
Comment (leplatrem) [inline README.md]: Will we switch when 67 becomes release? Maybe we'll have to wait a one or two more cycles to make sure our population becomes big enough?

File: README.md [hunk 3]
```diff
@@ -8,6 +8,62 @@ a threshold.
...
+appropriate) and still use this script to worry about the percentage thresholds.
+And the strategy for notifications should hopefully not change.
+
+Essentionally, once we switch to real-time Telemetry Events, nothing changes in
```
Comment (leplatrem) [inline README.md]: Essentially

File: README.md [hunk 4]
```diff
@@ -8,6 +8,62 @@ a threshold.
...
+terms of the surface API but the underlying data is more up-to-date and the response
+time of reacting to failure spikes is faster.
```
Comment (leplatrem) [inline README.md]: We could also mention that in the mid/long term, based on the Telemetry tools available to do some alerting we could decomission this one.

File: README.md [hunk 5]
```diff
@@ -8,6 +8,62 @@ a threshold.
...
+
+### Configuration
+
+This script encodes all configuration inside `main.py` as defaults for things
+that can be overwritten by environment variables.
+
+The only **mandatory environment variable** is `REDASH_API_KEY`. You get this
+by logging in [https://sql.telemetry.mozilla.org](https://sql.telemetry.mozilla.org)
+and click to edit the/any relevant query and in the upper right-hand corner there's a
+drop-down with the label "..." and inside it "Show API Key".
```
Comment (leplatrem) [inline README.md]: Maybe a screenshot would be relevant here?

File: README.md [hunk 6]
```diff
@@ -8,6 +8,62 @@ a threshold.
 The ultimately use case for this script is to run it periodically and use it to
 trigger alerts/notifications that the Product Delivery team can take heed of.
 
+## Architectural Overview
+
+### In Plain English
```
Comment (leplatrem) [inline README.md]: This title could imply that a diagram is coming next :) What do you think of *In Brief*?

File: README.md (hunk 1 above)
Comment (peterbe) [reply README.md]: I don't your point. `histogram_parent_uptake_remote_content_result_1` is hardcoded in the Redash query. So how is my sentence incorrect?

File: README.md (hunk 2 above)
Comment (peterbe) [reply README.md]: Days after 67 becomes release the numbers we currently get in https://sql.telemetry.mozilla.org/queries/61352 will quickly reduce to nothing, right?

Your patch to switch to Telemetry Events will *replace* the histogram stuff, no?

That reminds me, the change you landed in Nightly (which I can clearly see in `about:telemetry`) is already in effect. So, can I confirm that, with (a fork of) https://sql.telemetry.mozilla.org/queries/61352/source#158020, the numbers are all near 0 if I can filter that query down to just Nightlies?

File: README.md (hunk 3 above)
Comment (peterbe) [reply README.md]: Use the GitHub PR Suggestions button :)

File: README.md (hunk 5 above)
Comment (peterbe) [reply README.md]: Meh. Too much work and hassle for something 2-3 people are ever going to hack on :)

File: README.md (hunk 6 above)
Comment (peterbe) [reply README.md]: We can bikeshed and neither of us are native-tongue English speakers. I like "In Plain English" but I also don't care that much. 

The *intention* of this block is to answer, if a colleague asks, "What is this thing?"

File: README.md (hunk 2 above)
Comment (leplatrem) [reply README.md]: > Your patch to switch to Telemetry Events will _replace_ the histogram stuff, no?

Nope, it's both for now :)

Comment (peterbe) [reply README.md]: Wow! I totally missed that. Did the patch change after I first looked at it?
Either way, I'll make a note of this.

File: README.md [hunk 7]
```diff
@@ -8,6 +8,62 @@ a threshold.
...
+
+This is stateless script, written in Python, meant to be executed roughly once a day.
+It uses `requests.get()` to query [Redash](https://sql.telemetry.mozilla.org) for
```
[the remark this answers is not in the document]
Comment (peterbe) [reply README.md]: you're totally right!

[the remark this answers is not in the document]
Comment (peterbe) [reply README.md]: Can I say "make a `GET` request"?

[the remark this answers is not in the document]
Comment (glasserc) [reply README.md]: You could, but maybe just `It queries Redash ...` is good enough?

File: README.md [hunk 8]
```diff
@@ -8,6 +8,62 @@ a threshold.
...
+a query's data that is pre-emptively run every 24 hours. The data it analyses is a
+list of every collection stored in Remote Settings (Kinto) along with a count of
+every possible status (e.g. `up_to_date`, `network_error`, etc.)
+The script sums the "good" statuses and compares it against the "bad" statuses and
+if that ratio (expressed as a percentage) is over a certain threshold it alerts
+by sending an event to Sentry which notifies the team by email.
+
+### Historical Strategy
+
+As of Feb 2019, Firefox clients that use Remote Settings only send Telemetry events
```
[the remark this answers is not in the document]
Comment (peterbe) [reply README.md]: "pings" is a better and more accurate word.

File: README.md [hunk 9]
```diff
@@ -8,6 +8,62 @@ a threshold.
...
+in daily batches. I.e. it's _not_ real-time. The "uptake histogram" is buffered in
+the browser and will send periodically instead of as soon as possible. In Firefox
+Nightly (67 at the time of writing), we are switching to real-time Telemetry Events.
```
[the remark this answers is not in the document]
Comment (peterbe) [reply README.md]: @leplatrem Which bug would that be?

**Comment by peterbe** [comment]
@leplatrem Can you eye through it one last time? I don't want this to fester.

File: README.md (hunk 1 above)
Comment (leplatrem) [reply README.md]: The unified Uptake Telemetry is not specific to Remote Settings. Proof is that Shield and Normandy use it.

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (leplatrem) [reply README.md]: https://bugzilla.mozilla.org/show_bug.cgi?id=1517469
